packages/app/src/cli/services/function/build.ts

```diff
@@ -137,3 +156,122 @@ export async function runFunctionRunner(fun: ExtensionInstance, options: Functio
+    this.exports = exports
+  }
+
+  async bundle(fun: ExtensionInstance<FunctionConfigType>, options: JSFunctionBuildOptions) {
+    const javyFs = await findPathUp('node_modules/javy/fs', {
+      type: 'directory',
+      cwd: fun.directory,
+    })
+    if (!javyFs) {
+      throw new Error(
+        "Could not find the javy/fs runtime. Make sure you have 'javy/fs' installed",
+      )
+    }
+
+    if (!fun.entrySourceFilePath) {
+      throw new Error('Could not find your function entry point. It must be in src/index.js or src/index.ts')
+    }
+
+    const contents = this.entrypointContents
+    console.log(contents)
+
+    const esbuildOptions: Parameters<typeof esBuild>[0] = {
+      outfile: joinPath(fun.directory, 'dist/function.js'),
+      stdin: {
+        contents,
+        loader: 'ts',
+      },
+      alias: {
+        'user-function': fun.entrySourceFilePath,
+        'javy/fs': javyFs,
+      },
+      logLevel: 'silent',
+      bundle: true,
+      legalComments: 'none',
+      target: 'es2022',
+      format: 'esm',
+    }
+
+    return esBuild(esbuildOptions)
+  }
+
+  async compile(fun: ExtensionInstance<FunctionConfigType>, options: JSFunctionBuildOptions) {
+    const witPath = joinPath(fun.directory, 'dist',  'javy-world.wit')
+    await writeFile(witPath, this.wit)
+    const javyArgs = ['--wit', witPath, '-n', JAVY_WORLD]
+
+    return runJavy(fun, options, javyArgs)
+  }
+
+  get wit() {
+    // % escapes the name to avoid conflict with reserved words, if any
+    const witExports = this.exports.map((name) => `export %${hyphenate(name)}: func()`)
+    return `package function:impl
+
+world ${JAVY_WORLD} {
+  ${witExports.join('\n  ')}
+}`
+  }
+
+  get entrypointContents() {
+    const prelude = `
+import * as userExports from "user-function"
+import * as fs from "javy/fs"
+
+function run(fn) {
+  const input_data = fs.readFileSync(fs.STDIO.Stdin);
+  const input_str = new TextDecoder("utf-8").decode(input_data);
+  const input_obj = JSON.parse(input_str);
+  const output_obj = userfunction(input_obj);
+  const output_str = JSON.stringify(output_obj);
+  const output_data = new TextEncoder().encode(output_str);
+}`
+
+  const exports = this.exports.map((name) => {
+    const identifier = camelize("run_" + name);
+    return `
+function ${identifier}() { return run(userExports["${name}"]) }
+export { ${identifier} as "${hyphenate(name)}" }` })
+
+
+    return prelude + '\n' + exports.join("\n")
+  }
+}
+
+export function jsExports(fun: ExtensionInstance<FunctionConfigType>) {
+  const targets = fun.configuration.targeting || []
+  const withoutExport = targets.find((target) => !target.export)
+  const withExports = targets.filter((target) => Boolean(target.export))
+
+  if (withoutExport && withExports.length > 0) {
+    throw new Error(`Can't infer export name for '${withoutExport.target}'. All targets must have an export, or none.`)
+  }
+
+  return withExports.map((target) => target.export!)
+}
```

Comment:
> [..], the underlying reason is that we typically map default exports to _start in WASI. Explicitly accepting default as the name is confusing (should the behaviour diverge from the unnamed default that gets mapped to _start? what would people expect?)

Not exactly. `_start` is mapped to top-level evaluation of the JS and not invoking any functions. If we were to invoke a default exported arrow function as part of invoking `_start`, it would be a breaking change in Javy's existing behaviour. And also we wouldn't be able to support arguments in `_start`. I think we could support exporting a function named `default` for a default function export or default arrow function export. We just don't at the moment. See https://github.com/bytecodealliance/javy/issues/422.